File: cxx_cmake/lib/bowling_game/src/bowling_game.cpp
#include <iostream>
#include "bowling_game.hpp"

using namespace std;


bowling_game::bowling_game(){

}

bowling_game::~bowling_game(){

}

bool bowling_game::is_strike(int frame_index){
    return ( rolls[frame_index] == 10 );
}

bool bowling_game::is_spare(int frame_index){
    return ( (rolls[frame_index] + rolls[frame_index+1]) == 10 );
}

int bowling_game::strike_score(int frame_index){
    return 10 + rolls[frame_index+1] + rolls[frame_index+2];
}

int bowling_game::spare_score(int frame_index){
    return 10 + rolls[frame_index+2];
}

int bowling_game::nomal_score(int frame_index){
    return rolls[frame_index] + rolls[frame_index+1];
}

void bowling_game::roll(int pins){
    rolls[current_roll++] = pins;
}

int bowling_game::score(){
    int score = 0;
    int frame_index;
    int frame;
    frame_index=0;
    for(frame=0; frame<10; frame++){
        if( is_strike(frame_index) ){
            score += strike_score(frame_index);
            frame_index += 1;
        }else if( is_spare(frame_index) ){
            score += spare_score(frame_index);
            frame_index += 2;
        }else{
            score += nomal_score(frame_index);
            frame_index += 2;
        }
    }
    return score;
}

File: typescript_bun/bowling_game.ts
export class BowlingGame {
    private rolls: number[] = [];

    private isStrike(rollIndex: number): boolean {
        return this.rolls[rollIndex] === 10;
    }

    private isSpare(rollIndex: number): boolean {
        return this.rolls[rollIndex] + this.rolls[rollIndex + 1] === 10;
    }

    private strikeScore(frameIndex: number): number {
        return 10 + this.rolls[frameIndex + 1] + this.rolls[frameIndex + 2]
    }

    private spareScore(frameIndex: number): number {
        return 10 + this.rolls[frameIndex + 1]
    }

    roll(pins: number): void {
        this.rolls.push(pins);
    }

    score(): number {
        const maxIndex = this.rolls.length
        this.roll(0)
        this.roll(0)
        let score = 0;
        let frameIndex = 0;
        for (let frame = 0; frame < 10; frame++) {
            if (frameIndex >= maxIndex) {
                break;
            }
            if (this.isStrike(frameIndex)) {
                score += this.strikeScore(frameIndex)
                frameIndex += 1
            } else if (this.isSpare(frameIndex)) {
                score += this.spareScore(frameIndex)
                frameIndex += 2
            } else {
                score += this.rolls[frameIndex] + this.rolls[frameIndex + 1]
                frameIndex += 2
            }
        }
        return score;
    }
}
